### ISSUE TITLE
NVIDIA-KlausT ccminer does not work with 4060ti, 4090, or 3090

### ISSUE THREAD

**Issue by username-rand** [RainbowMiner/RainbowMiner#2467]
KlausT ccminer fails on 3090, 4060ti, and 4090 on HiveOS 18.04

Compiled with GCC 7.5 using the Nvidia CUDA Toolkit 11.6

[2023-09-12 13:25:17] Starting Stratum on stratum+tcp://neoscrypt.na.mine.zpool.ca:4233
[2023-09-12 13:25:17] 1 miner thread started, using 'neoscrypt' algorithm.
[2023-09-12 13:25:17] Stratum difficulty set to 128
[2023-09-12 13:25:20] GPU #1: using default intensity 19.000
free(): invalid pointer

so I built a ccminer KlausT from source (although the source doesn't contain XAYA whereas your version does) on ubuntu 18.04 and the CUDA Toolkit 11.7 and it works but doesn't have XAYA:

Compiled with GCC 7.5 using the Nvidia CUDA Toolkit 11.7

[2023-09-12 13:40:54] Starting Stratum on stratum+tcp://neoscrypt.mine.zergpool.com:4233
[2023-09-12 13:40:54] 1 miner thread started, using 'neoscrypt' algorithm.
[2023-09-12 13:40:55] Stratum difficulty set to 256
[2023-09-12 13:40:57] GPU #2: using default intensity 18.000
[2023-09-12 13:40:59] accepted: 1/1 (100.00%), 6170.95 kH/s yay!!!
[2023-09-12 13:41:01] accepted: 2/2 (100.00%), 6083.25 kH/s yay!!!
[2023-09-12 13:41:02] accepted: 3/3 (100.00%), 6080.25 kH/s yay!!!

Would you mind building a version of NVIDIA-KlausT with XAYA using CUDA Toolkit 11.7 or later so that it can be downloaded?  Or let me know where the source with XAYA is and I'll rebuild it myself.   KlausT was based on sp-hash which was based on tpruvot but code no longer looks anything like the current version of tpruvot.  KlausT final release was July 8, 2021.   So it is newer than tpruvot.  Here is ccminer based on tpruvot with XAYA and I can see the commit but it doesn't look easy to add to the KlausT code,

https://github.com/xaya/ccminer

So I currently am using my built version of KlausT but don't have XAYA like your version.  If you could build your version with XAYA so that I could download it for CUDA 11.7 or later or show me where to get the code, I have two workers that use 40xx GPU's.

I did notice that https://github.com/xaya/ccminer contains Allium and XAYA so I'm also trying to make this.  It also is based on tpruvot so is not as recent as KlausT but has both algos apparently.

Thanks

Rand

P.S. When I remade the KlausT ccminer I changed Makefile.am  to add CUDA 11 to:

nvcc_ARCH += -gencode=arch=compute_60,code=sm_60
nvcc_ARCH += -gencode=arch=compute_52,code=sm_52
nvcc_ARCH += -gencode=arch=compute_50,code=sm_50
#nvcc_ARCH += -gencode=arch=compute_37,code=sm_37
#nvcc_ARCH += -gencode=arch=compute_35,code=sm_35
nvcc_ARCH += -gencode=arch=compute_70,code=sm_70
nvcc_ARCH += -gencode=arch=compute_75,code=sm_75
nvcc_ARCH += -gencode=arch=compute_80,code=sm_80
nvcc_ARCH += -gencode=arch=compute_86,code=sm_86
nvcc_ARCH += -gencode=arch=compute_87,code=sm_87
nvcc_ARCH += -gencode=arch=compute_86,code=compute_86
#nvcc_ARCH += -gencode=arch=compute_89,code=sm_89
#nvcc_ARCH += -gencode=arch=compute_89,code=compute_89
#nvcc_ARCH += -gencode=arch=compute_90,code=sm_90
#nvcc_ARCH += -gencode=arch=compute_90,code=compute_90

Only change I made to source as I've not figured out how to add XAYA.  89 is used for CUDA 12 and I'd have to try building on ubuntu 20 and see if it might run on ubuntu 18.  Probably not but will try that later.  

P.P.S  Here is a link comparting arch with CUDA which is what I used
[](url)
https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

P.P.P.S  Benchmarked my KlausT vs [https://github.com/xaya/ccminer](url) after bulding xaya/ccminer.  KlausT is slightly faster for neoscrypt.  Could not benchmark xaya because the rebuilt KlausT doesn't have that algo.  Benchmarks were on 3060ti, 4060ti, and 4070.

KlausT 3060ti 1.74 MH/s  xaya/ccminer 1.6 MH/s
KlausT 4060ti 1.86 MH/s  xaya/ccminer 1.7 MH/s
KlausT 4070 3.096 MH/s  xaya/ccminer 3.87 MH/s

**Comment by RainbowMiner** [comment]
Sure, I'll re-compile it with a newer toolkit. Maybe we could already step up to CUDA 12. I'll try.

**Comment by username-rand** [comment]
Thanks, that would be very helpful and more efficient I think.  Currently I've built and running xaya/ccminer noted above and it appears to work built with Toolkit 11.7 on 3090, 4070, 3060ti, 4060ti, and 4090.  Your original build works fine on my 3080's.  Your KlausT version appears to be more efficient for neoscrypt.  Unable to test neoscrypt-xaya.  Adding CUDA 12 would be great, might make it  more efficient on the 40xx series.  I don't think it will be profitable, but I'm trying to figure out why crashing miners are crashing on my workers.  Sometimes it is OC.  Sometimes it is missing libraries on HiveOS (Ubuntu 18.04) such as Glibc2.29 required by some.  I'm at the C's :-).

Thanks again for your quick response and hard work!

Rand

**Comment by RainbowMiner** [comment]
Yes, I have integrated lots of optimization into that KlausT ccminer.

> Thanks again for your quick response and hard work!

My pleasure :) Thank you!

**Comment by username-rand** [comment]
Saw you had your ccminer repository.  Didn't see that before.  I've rebuilt  your version with CUDA 11.8 and running it until you have a new version to release.

Thanks

Rand

**Comment by BadWolf-63** [comment]
Hey, i don't know if you have dropped it totally or not, But i tinkered with your code a little, because my 3090 made the miner crash randomly while mining Xaya.
Found the problem (Misaligned memory) and fixed it, and while i was there i did a little code optimization of neoscrypt_gpu_hash_chacha1() and neoscrypt_gpu_hash_salsa1().
If you have time, give it a look.
[cuda_neoscrypt_tpruvot.zip](https://github.com/user-attachments/files/18419631/cuda_neoscrypt_tpruvot.zip)

**Comment by RainbowMiner** [comment]
Thank you! I will check recompile and try.
